### PR TITLE
Nodes per element

### PR DESCRIPTION
**Pull request by revengineering**
Fixes Issue #50, allowing the body of the foreachInit to have multiple HTML nodes per array element.

Also fixes a couple of minor browser compatibility issues with older versions of IE.

### CODE REVIEW

File: knockout-pre-rendered.js [hunk 1]
```diff
@@ -78,20 +78,34 @@
       // For e.g. <template> tags
       parentNode = sourceNode.content;
     } else if (sourceNode.tagName === 'SCRIPT') {
-      parentNode = sourceNode.innerHTML.match(/<tr[\s\S]*?<\/tr>/g) ? document.createElement('tbody') : document.createElement('div');
-      parentNode.innerHTML = sourceNode.innerHTML;
+      if(sourceNode.innerHTML.match(/<tr[\s\S]*?<\/tr>/g)) {
```
Comment (ErikSchierboom) [inline knockout-pre-rendered.js]: Very small nit: there should be a space after the `if`.

File: knockout-pre-rendered.js [hunk 2]
```diff
@@ -78,20 +78,34 @@
...
     }
 
     // Find the template and add it to the container
     var template = findFirstChild(parentNode, namedTemplate ? null : 'data-template');
-    container.insertBefore(template.cloneNode(true), null);
-
-    // Remove the template node
-    ko.removeNode(template);
 
+    for(var i = 0, node = template; i < nodesPerElement; i++) {
```
Comment (ErikSchierboom) [inline knockout-pre-rendered.js]: There should be a space after the `for`.

File: knockout-pre-rendered.js [hunk 3]
```diff
@@ -78,20 +78,34 @@
...
+      var currentNode = node;
+      container.insertBefore(node.cloneNode(true), null);
+      // Find next element sibling. (Some older browsers don't support nextElementSibling).
+      do { node = node.nextElementSibling || node.nextSibling; } while(node && node.nodeType !== 1);
```
Comment (ErikSchierboom) [inline knockout-pre-rendered.js]: I would prefer to put this on multiple lines:

```js
do {
  node = node.nextElementSibling || node.nextSibling; 
} while (node && node.nodeType !== 1); 
```

File: knockout-pre-rendered.js [hunk 4]
```diff
@@ -78,20 +78,34 @@
...
 
+    for(var i = 0, node = template; i < nodesPerElement; i++) {
+      var currentNode = node;
+      container.insertBefore(node.cloneNode(true), null);
+      // Find next element sibling. (Some older browsers don't support nextElementSibling).
+      do { node = node.nextElementSibling || node.nextSibling; } while(node && node.nodeType !== 1); 
+      // Remove current template node
+    if(deleteTemplateNodes) {
+        ko.removeNode(currentNode);
+    }
```
Comment (ErikSchierboom) [inline knockout-pre-rendered.js]: There appears to be some indentation problem here

**Comment by revengineering** [comment]
Should be all set.

**Comment by ErikSchierboom** [comment]
Merged! Thanks a lot @revengineering. I'll put out a new release tonight.

**Comment by revengineering** [comment]
My pleasure. Thanks for making this library! It has completely changed my perspective on knockout, and put it back in the running for my current and future projects.

**Comment by ErikSchierboom** [comment]
@revengineering That's great to hear!

I've just published version 0.9.2 to NPM and GitHub, which includes your changes. Again, thanks a lot!

**Comment by penguinstampede** [comment]
This update seems to break foreachInit if your looped elements (if that's the right term) are tables, so something like:

```
<div data-bind="foreachInit: { name: 'theTemplate', data: myData }">
<table data-init>
[usual table-y stuff here]
</table>
</div>
```

nets you the error `Cannot read property 'cloneNode' of null`

I wrote a test for this, and I fixed it by changing line 81:

`if (sourceNode.innerHTML.match(/<tr[\s\S]*?<\/tr>/g)) {`

to

`if (sourceNode.innerHTML.match(/<tr[\s\S]*?<\/tr>/g) && !sourceNode.innerHTML.match(/<table[\s\S]*?<\/table>/g)) {`

but I'm not sure if this is the best way to fix this?  I can make a new PR with the test/fix if you'd like, though.

**Comment by ErikSchierboom** [comment]
>  I can make a new PR with the test/fix if you'd like, though.

That would be great!

**Comment by revengineering** [comment]
The fix you proposed would work except if the tr elements contain any nested tables

I suggest an alternate regex of:

if (sourceNode.innerHTML.match(/^\s*<tr[\s\S]*?<\/tr>\s*$/i)

This way, it would only concern itself with templates where the first child is a tr element, regardless of whatever else might be inside them. (Also ignoring case, because browsers still understand uppercase tag names.)

**Comment by ErikSchierboom** [comment]
@penguinstampede The suggestion by @revengineering is probably good to follow, as the internals are better understood by @revengineering than by me currently :D

**Pull request merged**